File: mobile/modules/native-llm-runtime/ios/NativeLLMRuntimeModule.swift
import ExpoModulesCore
import Foundation

public class NativeLLMRuntimeModule: Module {
  private let backend: any LLMBackend = LlamaCppBackend()

  public func definition() -> ModuleDefinition {
    Name("NativeLLMRuntime")

    // MARK: - Health

    AsyncFunction("getLLMRuntimeHealth") { [self] () -> [String: Any] in
      var result: [String: Any] = [
        "available": backend.isLinked,
        "platform": "ios",
        "backend": backend.backendName,
        "supportsStreaming": backend.supportsStreaming,
        "supportsCancellation": backend.supportsCancellation,
        "supportsQuantizedModels": backend.supportsQuantizedModels,
        "supportedFormats": backend.supportedFormats,
      ]

      if let modelId = backend.currentModelId {
        result["loadedModelId"] = modelId
      } else {
        result["loadedModelId"] = NSNull()
      }

      if let reason = backend.reasonNotLinked() {
        result["reasonUnavailable"] = reason
      } else {
        result["reasonUnavailable"] = NSNull()
      }

      return result
    }

    // MARK: - Model discovery

    AsyncFunction("listInstalledModels") { () -> [[String: Any]] in
      return self.scanModelsDirectory()
    }

    // MARK: - Load / unload

    AsyncFunction("loadModel") { [self] (request: [String: Any]) -> [String: Any] in
      guard let modelId = request["modelId"] as? String,
            let localPath = request["localPath"] as? String
      else {
        throw NSError(
          domain: "NativeLLMRuntime",
          code: 400,
          userInfo: [NSLocalizedDescriptionKey: "modelId and localPath are required"]
        )
      }

      let ctxLen = (request["contextLength"] as? Int) ?? 2048
      let nativeReq = LoadModelRequestNative(
        modelId: modelId,
        localPath: localPath,
        preferredBackend: request["preferredBackend"] as? String,
        contextLength: ctxLen
      )

      do {
        let result = try await backend.loadModel(request: nativeReq)
        return [
          "loaded": result.loaded,
          "modelId": result.modelId,
          "backend": result.backend,
          "message": result.message,
        ]
      } catch let e as LlamaCppError {
        throw NSError(
          domain: "NativeLLMRuntime.LlamaCpp",
          code: 1,
          userInfo: [NSLocalizedDescriptionKey: e.errorDescription ?? e.localizedDescription]
        )
      }
    }

    AsyncFunction("unloadModel") { [self] (modelId: String) -> [String: Any] in
      let result = await backend.unloadModel(modelId: modelId)
      return [
        "unloaded": result.unloaded,
        "modelId": result.modelId,
        "message": result.message,
      ]
    }

    // MARK: - Inference (iOS-only in Phase 2B; Android stubs this at the TS layer)

    AsyncFunction("runInference") { [self] (request: [String: Any]) -> [String: Any] in
      guard let modelId = request["modelId"] as? String,
            let prompt = request["prompt"] as? String
      else {
        throw NSError(
          domain: "NativeLLMRuntime",
          code: 400,
          userInfo: [NSLocalizedDescriptionKey: "modelId and prompt are required"]
        )
      }

      let nativeReq = RunInferenceRequestNative(
        modelId: modelId,
        prompt: prompt,
        maxTokens: (request["maxTokens"] as? Int) ?? 256,
        temperature: (request["temperature"] as? Float) ?? 0.8,
        topP: (request["topP"] as? Float) ?? 0.95,
        stopSequences: (request["stopSequences"] as? [String]) ?? []
      )

      do {
        let result = try await backend.runInference(request: nativeReq)
        return [
          "text": result.text,
          "tokensGenerated": result.tokensGenerated,
          "tokensSeen": result.tokensSeen,
          "backend": result.backend,
          "modelId": result.modelId,
        ]
      } catch let e as LlamaCppError {
        throw NSError(
          domain: "NativeLLMRuntime.LlamaCpp",
          code: 2,
          userInfo: [NSLocalizedDescriptionKey: e.errorDescription ?? e.localizedDescription]
        )
      }
    }
  }

  // MARK: - Model file scanning

  private func scanModelsDirectory() -> [[String: Any]] {
    let fileManager = FileManager.default
    guard let documentsURL = fileManager.urls(for: .documentDirectory, in: .userDomainMask).first
    else { return [] }

    let fileExtensions: Set<String> = ["gguf", "bin"]
    let packageExtensions: Set<String> = ["mlmodelc", "mlpackage"]
    var models: [[String: Any]] = []
    let iso = ISO8601DateFormatter()

    guard let enumerator = fileManager.enumerator(
      at: documentsURL,
      includingPropertiesForKeys: [.fileSizeKey, .creationDateKey, .isDirectoryKey],
      options: [.skipsHiddenFiles]
    ) else { return [] }

    for case let url as URL in enumerator {
      let ext = url.pathExtension.lowercased()
      var isDir: ObjCBool = false
      fileManager.fileExists(atPath: url.path, isDirectory: &isDir)

      if isDir.boolValue {
        if packageExtensions.contains(ext) {
          enumerator.skipDescendants()
          let attrs = try? url.resourceValues(forKeys: [.creationDateKey])
          let discoveredAt = attrs?.creationDate.map { iso.string(from: $0) } ?? ""
          models.append([
            "id": url.lastPathComponent,
            "name": url.deletingPathExtension().lastPathComponent,
            "localPath": url.path,
            "format": ext,
            "sizeBytes": 0,
            "discoveredAt": discoveredAt,
          ])
        }
        continue
      }

      guard fileExtensions.contains(ext) else { continue }
      let attrs = try? url.resourceValues(forKeys: [.fileSizeKey, .creationDateKey])
      let sizeBytes = attrs?.fileSize ?? 0
      let discoveredAt = attrs?.creationDate.map { iso.string(from: $0) } ?? ""
      let format: String
      switch ext {
      case "gguf": format = "gguf"
      case "bin":  format = "bin"
      default:     format = "unknown"
      }
      models.append([
        "id": url.lastPathComponent,
        "name": url.deletingPathExtension().lastPathComponent,
        "localPath": url.path,
        "format": format,
        "sizeBytes": sizeBytes,
        "discoveredAt": discoveredAt,
      ])
    }

    return models
  }
}

File: mobile/modules/native-llm-runtime/android/src/main/java/expo/modules/nativellmruntime/NativeLLMRuntimeModule.kt
package expo.modules.nativellmruntime

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.io.File
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

class NativeLLMRuntimeModule : Module() {
  private var loadedModelId: String? = null

  override fun definition() = ModuleDefinition {
    Name("NativeLLMRuntime")

    AsyncFunction("getLLMRuntimeHealth") {
      val loadedId = loadedModelId
      mapOf(
        "available" to false,
        "isLinked" to false,
        "platform" to "android",
        "backend" to "none",
        "supportsStreaming" to false,
        "supportsCancellation" to false,
        "supportsQuantizedModels" to false,
        "supportedFormats" to emptyList<String>(),
        "loadedModelId" to loadedId,
        "reasonUnavailable" to "No Android inference backend linked yet. iOS llama.cpp exists; Android backend is planned for a later phase.",
      )
    }

    AsyncFunction("listInstalledModels") {
      scanModelsDirectory()
    }

    AsyncFunction("loadModel") { request: Map<String, Any?> ->
      val modelId = request["modelId"] as? String
        ?: throw IllegalArgumentException("modelId is required")
      val localPath = request["localPath"] as? String
        ?: throw IllegalArgumentException("localPath is required")

      val file = File(localPath)
      if (!file.exists()) {
        return@AsyncFunction mapOf(
          "loaded" to false,
          "modelId" to modelId,
          "backend" to "none",
          "message" to "File not found at path: $localPath",
        )
      }

      mapOf(
        "loaded" to false,
        "modelId" to modelId,
        "backend" to "none",
        "message" to "File validated, but Android inference backend is not linked yet.",
      )
    }

    AsyncFunction("unloadModel") { modelId: String ->
      val wasLoaded = loadedModelId == modelId
      loadedModelId = null
      mapOf(
        "unloaded" to wasLoaded,
        "modelId" to modelId,
        "message" to if (wasLoaded) "Model state cleared." else "No model was loaded with that ID.",
      )
    }
  }

  private fun scanModelsDirectory(): List<Map<String, Any?>> {
    val context = appContext.reactContext ?: return emptyList()
    val filesDir = context.filesDir
    val dateFormat = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", Locale.US)
    val modelExtensions = setOf("gguf", "bin")
    val models = mutableListOf<Map<String, Any?>>()

    filesDir.walkTopDown().filter { it.isFile }.forEach { file ->
      val ext = file.extension.lowercase(Locale.US)
      if (ext !in modelExtensions) return@forEach

      val format = when (ext) {
        "gguf" -> "gguf"
        "bin" -> "bin"
        else -> "unknown"
      }

      models.add(
        mapOf(
          "id" to file.name,
          "name" to file.nameWithoutExtension,
          "localPath" to file.absolutePath,
          "format" to format,
          "sizeBytes" to file.length(),
          "discoveredAt" to dateFormat.format(Date(file.lastModified())),
        )
      )
    }

    return models
  }
}
